src/Nethermind/Nethermind.Runner/Program.cs

```diff
@@ -41,557 +42,522 @@
-                Environment.ExitCode = withExit.ExitCode;
+                plugins.Add(plugin);
             }
-            else
-            {
-                Environment.ExitCode = ExitCodes.GeneralError;
-            }
-            throw;
         }
-        finally
+        catch (Exception ex)
         {
-            _appClosed.Wait();
+            if (logger.IsError) logger.Error($"Failed to create plugin {pluginType.FullName}", ex);
         }
     }
 
-    [GeneratedRegex("^Unexpected value '(?<Value>.+)' for option '(?<Name>.+)'", RegexOptions.Singleline)]
-    private static partial Regex GetUnexpectedConfigValueRegex();
+    INethermindApi nethermindApi = apiBuilder.Create(plugins.OfType<IConsensusPlugin>());
+    ((List<INethermindPlugin>)nethermindApi.Plugins).AddRange(plugins);
+    nethermindApi.ProcessExit = processExitSource;
 
-    private static IntPtr OnResolvingUnmanagedDll(Assembly _, string nativeLibraryName)
+    EthereumRunner ethereumRunner = new(nethermindApi);
+    try
     {
-        var alternativePath = nativeLibraryName switch
-        {
-            "libdl" => "libdl.so.2",
-            _ => null
-        };
-
-        return alternativePath is null ? IntPtr.Zero : NativeLibrary.Load(alternativePath);
+        await ethereumRunner.Start(processExitSource.Token);
+        await processExitSource.ExitTask;
+    }
+    catch (OperationCanceledException)
+    {
+        if (logger.IsTrace) logger.Trace("Nethermind operation was canceled.");
     }
+    catch (Exception ex)
+    {
+        if (logger.IsError) logger.Error(unhandledError, ex);
+        processExitSource.Exit(ex is IExceptionWithExitCode withExit ? withExit.ExitCode : ExitCodes.GeneralError);
+    }
+
+    logger.Info("Nethermind is shutting down... Please wait until all activities are stopped.");
+
+    await ethereumRunner.StopAsync();
+
+    logger.Info("Nethermind is shut down");
+
+    exit.Set();
+
+    return processExitSource.ExitCode;
+}
+
+void AddConfigurationOptions(CliCommand command)
+{
+    IEnumerable<Type> configTypes = TypeDiscovery
+        .FindNethermindBasedTypes(typeof(IConfig))
+        .Where(ct => ct.IsInterface);
 
-    private static void BuildOptionsFromConfigFiles(CommandLineApplication app)
+    foreach (Type configType in
+        configTypes.Where(ct => !ct.IsAssignableTo(typeof(INoCategoryConfig))).OrderBy(c => c.Name))
     {
-        Type configurationType = typeof(IConfig);
-        IEnumerable<Type> configTypes = TypeDiscovery.FindNethermindBasedTypes(configurationType)
-            .Where(ct => ct.IsInterface);
+        if (configType is null)
+        {
+            continue;
+        }
+
+        ConfigCategoryAttribute? typeLevel = configType.GetCustomAttribute<ConfigCategoryAttribute>();
 
-        foreach (Type configType in configTypes.Where(ct => !ct.IsAssignableTo(typeof(INoCategoryConfig))).OrderBy(c => c.Name))
+        if (typeLevel is not null && (typeLevel.DisabledForCli || typeLevel.HiddenFromDocs))
```

Comment:
`HiddenFromDocs` here won't disable them?

Reply:
It will. The vast majority of our undocumented CLI options don't have descriptions, so it doesn't make sense to list them with `--help`. They look cryptic.